tortoisebeacon/weakcoin/handler.go

```diff
@@ -0,0 +1,67 @@
+package weakcoin
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+)
+
+// HandleSerializedMessage defines method to handle Tortoise Beacon Weak Coin Messages from gossip.
+func (wc *WeakCoin) HandleSerializedMessage(ctx context.Context, data service.GossipMessage, sync service.Fetcher) {
+	wc.logger.With().Debug("received weak coin message",
+		log.String("from", data.Sender().String()),
+		log.Binary("data", data.Bytes()),
+	)
+
+	var message Message
+	if err := types.BytesToInterface(data.Bytes(), &message); err != nil {
+		wc.logger.With().Debug("received invalid weak coin message",
+			log.Binary("message", data.Bytes()),
+			log.Err(err))
+		return
+	}
+
+	if err := wc.receiveMessage(message); err != nil {
+		wc.logger.With().Debug("received invalid proposal",
+			log.Err(err),
+			log.Uint32("epoch_id", uint32(message.Epoch)),
+			log.Uint64("round_id", uint64(message.Round)))
+		return
+	}
+	data.ReportValidation(ctx, GossipProtocol)
+}
+
+func (wc *WeakCoin) receiveMessage(message Message) error {
+	if wc.aboveThreshold(message.Signature) {
+		return fmt.Errorf("proposal %x is above threshold", message.Signature)
+	}
+
+	wc.mu.Lock()
+	defer wc.mu.Unlock()
+	if wc.epoch != message.Epoch || wc.round != message.Round {
+		if wc.isNextRound(message.Epoch, message.Round) && len(wc.nextRoundBuffer) < cap(wc.nextRoundBuffer) {
+			wc.nextRoundBuffer = append(wc.nextRoundBuffer, message)
+			return nil
+		}
+		return fmt.Errorf("message for the wrong round %v/%v", message.Epoch, message.Round)
+	}
+	return wc.updateProposal(message)
+}
+
+func (wc *WeakCoin) isNextRound(epoch types.EpochID, round types.RoundID) bool {
```

Comment:
What do you think of adding a test for this method? There're some calculations here and a test could help ensure that they are correct

Reply:
this method is tested. the coverage is close to 100. creating unit tests for non public methods is not a good practice and should be avoided.